numpyro/infer/util.py

```diff
@@ -68,6 +69,49 @@ def log_density(model, model_args, model_kwargs, params):
     return log_joint, model_trace
 
 
+class _stop_gradient_at_nonreparameterized(Messenger):
+    """
+    Stop gradient for samples at latent sample sites that has_rsample=False.
+    """
+
+    def postprocess_message(self, msg):
+        if (
+            msg["type"] == "sample"
+            and (not msg["is_observed"])
+            and (not msg["fn"].has_rsample)
+        ):
+            msg["value"] = lax.stop_gradient(msg["value"])
```

Comment:
We can distinguish `sample` and `rsample` in Distribution but that might require many changes. So I made this handler to stop gradient for samples at sites with `has_rsample=False`.

Reply:
what are `intermediates` here and elsewhere?

Reply:
Those are intermediate values of transformed distributions, performed [here](https://github.com/pytorch/pytorch/blob/7949fdd2b64faef1d90510784cb57190dcf87c69/torch/distributions/transformed_distribution.py#L117). In Pyro, they are cached in transforms class. In NumPyro, we store them in `intermediates` field. We stop gradient for them to match PyTorch behavior (but I'm not sure if PyTorch behavior is correct - i.e. I'm not sure if we need to stop gradient for intermediate values).

Reply:
@fritzo I think that you made some changes in [this PR](https://github.com/pytorch/pytorch/pull/5086/files#diff-8b108de0fae81bfc4f939e40322885653fc3c957b2f8ad0ff25aafc16bf2c639L56-R60). What do you think?

Reply:
> What do you think?

Exactly what question are you asking?  I believe [the PR you mentioned](https://github.com/pytorch/pytorch/pull/5086/files#diff-8b108de0fae81bfc4f939e40322885653fc3c957b2f8ad0ff25aafc16bf2c639L56-R60) ensures that Pyro's `TransformedDistribution.sample()` is not accidentally reparametrized when the base dist is nonreparametrized but the transforms are differentiable.  I don't know enough about this PR #1044 or NumPyro's rsample design to say any more.

Reply:
Let me give an example. Assume the distribution that we are interested in is: `TransformedDistribution(FakeNormal(0, 1), [AffineTransform(0, scale), AffineTransform(0, 2)])`, where `scale` is a learnable parameter. To draw a sample of that distribution, currently in Pyro, we return
```
x = FakeNormal().sample().detach()
y = 2 * scale.detach() * x
return y
```
Before that PyTorch PR, we return `2 * scale * x`.

I guess we are interested in `log_prob(y)`, which is calculated by
```python
FakeNormal().log_prob(x) + log(2) + log(scale)
```
in both situations. Now, assume that, `AffineTransform(0, scale).log_abs_det_jacobian(x, y)` is implemented by `return log(y / x)` rather than `return scale`. Then `log_prob(y)` will be calculated by
```
FakeNormal().log_prob(x) + log(2) + log(scale)
```
if `y = 2 * scale * x`, and will be calculated by
```
FakeNormal().log_prob(x) + log(2) + log(scale.detach())
```
if `y = 2 * scale.detach() * x`. Now two implementations give two different gradients w.r.t. `scale` parameter. I was a bit confused about that. _I'm not sure_ if that change is intended or not. That's why I asked. It is not relevant to NumPyro implementation.

Reply:
well if you do find a bug in `.log_abs_det_jacobian()` then https://github.com/pytorch/pytorch/issues/new seems like a good place for extended discussion

Reply:
Hmm, I didn't expect to find a bug. I thought there would be a simple interpretation here. Let me think about it more. Creating a PyTorch issue is not necessary because I couldn't find a real transform that has the pattern `Transform(some_params).logdet(x,y) = y/x`. Let's discuss in some lunch instead. :D 
@martinjankowiak How about we defer this discussion to the future? I don't actually understand much about those different interpretations.